### PR TITLE
quick review

### PR DESCRIPTION
**Pull request by hatskier**

### CODE REVIEW

File: src/gateway/tasks/loadPeers.ts [hunk 1]
```diff
@@ -16,6 +16,9 @@ export async function runLoadPeersTask(context: Application.BaseContext) {
 
   logger.info("Starting [loadPeers] task.");
 
+  // Hmm, why not use `setInterval` instead?
```
Comment (ppedziwiatr) [inline src/gateway/tasks/loadPeers.ts:19]: the reason is described here - https://javascript.info/settimeout-setinterval#nested-settimeout
and here https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Timeouts_and_intervals#recursive_timeouts .

I want to have a fixed delay between consecutive calls to arweave.net gateway and peers.

File: src/gateway/tasks/loadPeers.ts [hunk 2]
```diff
@@ -37,26 +40,28 @@ export async function loadPeers(context: Application.BaseContext) {
   logger.info("Updating peers...");
 
   let newPeers: PeerList = [];
+  // Do we really need this try cactch?
+  // In case of errors it will end this functions gracefully and
+  // will print "Peers check complete" (which is not true)
```
Comment (ppedziwiatr) [inline src/gateway/tasks/loadPeers.ts:45]: yup, good point, forgot I'm catching in .then().catch() :-)

File: src/gateway/tasks/loadPeers.ts [hunk 3]
```diff
@@ -37,26 +40,28 @@ export async function loadPeers(context: Application.BaseContext) {
...
-      return currentPeer.peer === peer
-    })) {
+    if (!newPeers.find(peer => currentPeer.peer === peer)) {
       peersToRemove.push(currentPeer.peer);
     }
   });
 
   logger.debug("Removing no longer available peers", peersToRemove);
 
+  // I'd rename removed to removedCount. What do you think?
```
Comment (ppedziwiatr) [inline src/gateway/tasks/loadPeers.ts:64]: 👍

File: src/gateway/tasks/loadPeers.ts [hunk 4]
```diff
@@ -66,6 +71,8 @@ export async function loadPeers(context: Application.BaseContext) {
   for (const peer of newPeers) {
     logger.debug(`Checking Arweave peer ${peer} [${(newPeers.indexOf(peer) + 1)} / ${newPeers.length}]`);
     try {
+      // For me it's a bit misleading that you use Benchmark module from the redstone-smartweave module
+      // Maybe we can create a separate NPM module for benchmarking or check if there are ready solutions
```
Comment (ppedziwiatr) [inline src/gateway/tasks/loadPeers.ts:75]: yup, that's the plan for the nearest future - to move both the logger and  benchmark to a separate utility lib...

File: src/gateway/tasks/verifyInteractions.ts [hunk 5]
```diff
@@ -17,8 +17,13 @@ type ConfirmationRoundResult = {
   confirmations: number;
 }[];
 
+// TODO: in future we could unify the tasks interface and logic in a BaseTask.ts class (for example)
+// Now it looks like some pieces of code are duplicated (like repeated executions)
```
Comment (ppedziwiatr) [inline src/gateway/tasks/verifyInteractions.ts:21]: 👍

File: src/gateway/tasks/verifyInteractions.ts [hunk 6]
```diff
@@ -17,8 +17,13 @@ type ConfirmationRoundResult = {
...
+
 export async function runVerifyInteractionsTask(context: Application.BaseContext) {
   await verifyInteractions(context);
+  // The same question: why not to use setInterval instead
```
Comment (ppedziwiatr) [inline src/gateway/tasks/verifyInteractions.ts:25]: same answer ;-) ...I'll observe the memory usage..

File: src/gateway/runGateway.ts [hunk 7]
```diff
@@ -4,7 +4,8 @@ import {runLoadPeersTask} from "./tasks/loadPeers";
 import {runSyncBlocksTask} from "./tasks/syncBlocks";
 import {runVerifyInteractionsTask} from "./tasks/verifyInteractions";
 
-
+// TODO: I would move the `initGatewayDb` to a seaparate module (e.g. schemas.ts or db.ts)
+// along with the `INTERACTIONS_TABLE` type
 export type INTERACTIONS_TABLE = {
```
Comment (jakub-wojciechowski) [inline src/gateway/runGateway.ts:9]: Probably I'm missing sth, but don't we need to store contractIds for interactWrites to make faster queries?

File: src/gateway/runGateway.ts [hunk 8]
```diff
@@ -70,7 +70,7 @@ export async function initGatewayDb(db: Knex) {
  * TX_CONFIRMATION_SUCCESSFUL_ROUNDS query rounds (to randomly selected at each round peers).
```
Comment (jakub-wojciechowski) [inline src/gateway/runGateway.ts:70]: Sorry if it's a silly idea but have you consider deciding on the transaction confirmation status by checking the confirmation status of the block (hash) in which the transaction was included. Potential benefits:
1) Fewer roundtrips for multiple smart contract transaction in single block
2) Possible faster answers about block status as it's more widely used parameter crucial for building new blocks

File: src/gateway/tasks/loadPeers.ts (hunk 1 above)
Comment (hatskier) [reply src/gateway/tasks/loadPeers.ts:19]: Bbetween calls completion, right?

File: src/gateway/tasks/syncBlocks.ts [hunk 9]
```diff
@@ -63,6 +63,9 @@ interface ReqVariables {
 
 
 export async function runSyncBlocksTask(context: Application.BaseContext) {
+  // Hmm, not sure if this approach is much better, as it can result in fastly increasing memory usage
```
Comment (jakub-wojciechowski) [inline src/gateway/tasks/syncBlocks.ts:66]: It's not a priority and we probably shouldn't implement it now. What do you guys think about abstracting the execution in a job queue, like https://github.com/bee-queue/bee-queue

1) Pros -> better control over concurrency, memory, debugging, (most importantly being able to gracefully continue after a shutdown)
2) Cons -> added complexity, too much enterprise patterns, additional components (redis)

File: src/gateway/tasks/loadPeers.ts (hunk 1 above)
Comment (ppedziwiatr) [reply src/gateway/tasks/loadPeers.ts:19]: yup, exactly. As the response times from both the gateway and peers fluctuate heavily I want to be sure that there's at least some time between the calls (in order to not get rate limitted, etc.)

File: src/gateway/runGateway.ts (hunk 8 above)
Comment (ppedziwiatr) [reply src/gateway/runGateway.ts:70]: All the orphaned transactions for our loot contract come from one block - https://d1o5nlqr4okus2.cloudfront.net/gateway/interactions?contractId=Daj-MNSnH55TDfxqC7v4eq0lKzVIwh98srUaWqyuZtY&confirmationStatus=orphaned , blockId `bAUsEgzlbGyhJpyg-klaqw9GXhrqiVpbHsV2tjpGJ02faWGzLtuisqGagTqk5tQS`

Yet this block is normally visible on Arweave - https://arweave.net/block/hash/bAUsEgzlbGyhJpyg-klaqw9GXhrqiVpbHsV2tjpGJ02faWGzLtuisqGagTqk5tQS
and viewblock - https://viewblock.io/arweave/block/769686

How exactly this block-based verification should work?

File: src/gateway/runGateway.ts (hunk 7 above)
Comment (ppedziwiatr) [reply src/gateway/runGateway.ts:9]: the only thing that is missing here is the internal writes handling ;-), I'm aware of that, planned for the version 0.0.2 ;-)

File: src/gateway/tasks/verifyInteractions.ts [hunk 10]
```diff
@@ -244,7 +255,7 @@ async function verifyInteractions(context: Application.BaseContext) {
         } else {
           logger.warn("Different response from peers for", {
             current_peer: statusesRounds[j][i],
-            prev_peer: statusesRounds[j - 1][i]
+            prev_peer: statusesRounds[j - 1][i] // Will it work correctly for j == 0?
```
Comment (ppedziwiatr) [inline src/gateway/tasks/verifyInteractions.ts:258]: obviously it would not work properly for j = 0, but if we're in this branch (which means that we have different responses from rounds) than j >= 1.
i'll add a comment in code.

File: src/gateway/tasks/syncBlocks.ts (hunk 9 above)
Comment (ppedziwiatr) [reply src/gateway/tasks/syncBlocks.ts:66]: re. greacefully continuing after a shutdown - currently db gives us such feature..we're simply searching for the last processed block height at (after starting the gateway again) continue the processing..same in case of verifying interactions...
One place where it could be useful that I can think of is:
One process puts new blocks/interactions into queue (queueueueue https://www.youtube.com/watch?v=0V9Ua538jMI&t=223s ;-)) and there are multiple process/nodes that take values from the queue in order to verify them. Currently it's a bit hard to introduce such parallelism, because each node would start to verify the transactions at (roughly) the exact same point (so they would basically overlap their work) - I had to introduce some manual block height offsets in the sql queries when I was in the process of verifying all the interactions by a set of 12 nodes.

File: src/gateway/runGateway.ts (hunk 8 above)
Comment (ppedziwiatr) [reply src/gateway/runGateway.ts:70]: I've made some more research...it seems that the "orphaned" transactions issue does not come from the forked blocks...it looks like the arweave gateway caches the transactions that are gossiped for the next block, but does not verify if they are finally becoming part of this block (?)

From the above example:
 1. block height `769686`:
 - indep_hash `bAUsEgzlbGyhJpyg-klaqw9GXhrqiVpbHsV2tjpGJ02faWGzLtuisqGagTqk5tQS`, - https://arweave.net/block/hash/bAUsEgzlbGyhJpyg-klaqw9GXhrqiVpbHsV2tjpGJ02faWGzLtuisqGagTqk5tQS - https://viewblock.io/arweave/block/769686
 
 2. block height `769687`:
 - indep_hash `1SWZh9jw23JMCn0sfLCCwtZ7L9Xb3wkSWJO7MbXy7FLieyyjkz14LcnDEWobvu31`, 
 - previous_block `bAUsEgzlbGyhJpyg-klaqw9GXhrqiVpbHsV2tjpGJ02faWGzLtuisqGagTqk5tQS`
 - https://arweave.net/block/hash/1SWZh9jw23JMCn0sfLCCwtZ7L9Xb3wkSWJO7MbXy7FLieyyjkz14LcnDEWobvu31
 - https://viewblock.io/arweave/block/769687
 
 3. block height `769688`:
 - indep_hash `-QZNOuZyQj27rdGT3SfDp2I8xayG81eyLd7bGtv03xHiBfbu_v93BXUTfwxgyoCr`,
 - previous_block `1SWZh9jw23JMCn0sfLCCwtZ7L9Xb3wkSWJO7MbXy7FLieyyjkz14LcnDEWobvu31`
 - https://arweave.net/block/hash/-QZNOuZyQj27rdGT3SfDp2I8xayG81eyLd7bGtv03xHiBfbu_v93BXUTfwxgyoCr
 - https://viewblock.io/arweave/block/769688
 
 4. block height `769689`:
 - indep_hash `7s2pOt_0wAWfK3x4Tqk2Mro0DEpRwwquTL59BeW_o3gQzrsB3-ZF8RcfpVaWu5qs`
 - previous_block `-QZNOuZyQj27rdGT3SfDp2I8xayG81eyLd7bGtv03xHiBfbu_v93BXUTfwxgyoCr`
 - https://arweave.net/block/hash/7s2pOt_0wAWfK3x4Tqk2Mro0DEpRwwquTL59BeW_o3gQzrsB3-ZF8RcfpVaWu5qs
 - https://viewblock.io/arweave/block/769689
 
 5. block height `769690`:
 - indep_hash `mj52-sl7iOdNFFkkxqyQ9WTPjxwmy4I6rp3ZW81wKKIgdpwTqCh79g1a0N8qYe17`
 - previous_block `7s2pOt_0wAWfK3x4Tqk2Mro0DEpRwwquTL59BeW_o3gQzrsB3-ZF8RcfpVaWu5qs`
 - https://arweave.net/block/hash/mj52-sl7iOdNFFkkxqyQ9WTPjxwmy4I6rp3ZW81wKKIgdpwTqCh79g1a0N8qYe17
 - https://viewblock.io/arweave/block/769690
 
 ...and so on...TBH I don't think that all of these blocks are part of the fork...
 
 Though there's probably a bit more efficient way to verify transactions that have just come to my mind:
 1. load a bunch of `non_confimed` interactions and group them by block indep_hash
 2. load blocks data for all the blocks from point 1 (using the `https://arweave.net/block/hash/<indep_hash>` endpoint
 3. verify if the transactions are within the `txs` field from the response from point 2...
 
In case of our loot contract - none of the orphaned transactions are on the `txs` list of the `bAUsEgzlbGyhJpyg-klaqw9GXhrqiVpbHsV2tjpGJ02faWGzLtuisqGagTqk5tQS`.
 
image

This could potentially generate much less requests...

To sum up - I believe that "orhpaned" in this case means "transactions that are not within any block (but for some reason were cached by the gateway" and does NOT mean "transactions that are part of a forked block".

File: src/gateway/runGateway.ts (hunk 7 above)
Comment (ppedziwiatr) [reply src/gateway/runGateway.ts:9]: https://github.com/redstone-finance/redstone-sw-gateway/issues/15

File: src/gateway/tasks/syncBlocks.ts (hunk 9 above)
Comment (ppedziwiatr) [reply src/gateway/tasks/syncBlocks.ts:66]: let's move the conversation to https://github.com/redstone-finance/redstone-sw-gateway/issues/16

File: src/gateway/init.ts [hunk 11]
```diff
@@ -15,6 +15,9 @@ const envPath = argv.env_path || '.secrets/.env';
 
 const cors = require('@koa/cors');
 
+// TODO: why do we delcare a module for an external dependency
+// If there are no types, I'd put it in decs.d.ts
+// More info: https://medium.com/@steveruiz/using-a-javascript-library-without-type-declarations-in-a-typescript-project-3643490015f3
```
Comment (ppedziwiatr) [inline src/gateway/init.ts:20]: moved to generics.

File: src/gateway/runGateway.ts [hunk 12]
```diff
@@ -4,7 +4,8 @@ import {runLoadPeersTask} from "./tasks/loadPeers";
 import {runSyncBlocksTask} from "./tasks/syncBlocks";
 import {runVerifyInteractionsTask} from "./tasks/verifyInteractions";
 
-
+// TODO: I would move the `initGatewayDb` to a seaparate module (e.g. schemas.ts or db.ts)
```
Comment (ppedziwiatr) [inline src/gateway/runGateway.ts:7]: ok

File: src/gateway/runGateway.ts [hunk 13]
```diff
@@ -70,7 +70,7 @@ export async function initGatewayDb(db: Knex) {
  * TX_CONFIRMATION_SUCCESSFUL_ROUNDS query rounds (to randomly selected at each round peers).
  * Only if we get TX_CONFIRMATION_SUCCESSFUL_ROUNDS within TX_CONFIRMATION_MAX_ROUNDS
  * AND response for the given transaction is the same for all the successful rounds
- * - the "confirmation" info for given transaction in updated in the the database.
+ * - the "confirmation" info for given transaction in (in -> is?) updated in the the database.
```
Comment (ppedziwiatr) [inline src/gateway/runGateway.ts:73]: yup

File: src/gateway/tasks/loadPeers.ts [hunk 14]
```diff
@@ -37,26 +40,28 @@ export async function loadPeers(context: Application.BaseContext) {
...
   let newPeers: PeerList = [];
+  // Do we really need this try cactch?
+  // In case of errors it will end this functions gracefully and
+  // will print "Peers check complete" (which is not true)
   try {
     newPeers = await arweave.network.getPeers();
   } catch (e) {
     logger.error("Error from Arweave while loading peers", e);
-    return;
+    return; // can we throw e here instead of calling return
```
Comment (ppedziwiatr) [inline src/gateway/tasks/loadPeers.ts:50]: removed the try catch
